File: src/csharp/pulse/cdm/scalars/SEScalarLengthPerTime.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class LengthPerTimeUnit : Unit
  {
    private LengthPerTimeUnit(string v) { Value = v; }

    public static LengthPerTimeUnit m_Per_s = new LengthPerTimeUnit("m/s");
    public static LengthPerTimeUnit ft_Per_s = new LengthPerTimeUnit("ft/s");

    public static LengthPerTimeUnit FromString(string u)
    {
      if (u == LengthPerTimeUnit.m_Per_s.ToString())
        return LengthPerTimeUnit.m_Per_s;
      if (u == LengthPerTimeUnit.ft_Per_s.ToString())
        return LengthPerTimeUnit.ft_Per_s;

      throw new System.ArgumentException(u + " is not a know length per time unit");
    }
  }


  public class SEScalarLengthPerTime : SEScalar
  {
    protected LengthPerTimeUnit unit;

    public SEScalarLengthPerTime() : base()
    {

    }

    public SEScalarLengthPerTime(SEScalarLengthPerTime from)
    {
      Set(from);
    }

    public SEScalarLengthPerTime(double value, LengthPerTimeUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarLengthPerTime from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }


    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, LengthPerTimeUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }


    public double GetValue(LengthPerTimeUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public LengthPerTimeUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarLengthPerTime.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarLengthPerTime.h"

const LengthPerTimeUnit LengthPerTimeUnit::m_Per_s("m/s");
const LengthPerTimeUnit LengthPerTimeUnit::cm_Per_s("cm/s");
const LengthPerTimeUnit LengthPerTimeUnit::m_Per_min("m/min");
const LengthPerTimeUnit LengthPerTimeUnit::cm_Per_min("cm/min");
const LengthPerTimeUnit LengthPerTimeUnit::ft_Per_s("ft/s");
const LengthPerTimeUnit LengthPerTimeUnit::ft_Per_min("ft/min");
const LengthPerTimeUnit LengthPerTimeUnit::km_Per_hr("km/hr");

bool LengthPerTimeUnit::IsValidUnit(const std::string& unit)
{
  if (m_Per_s.GetString().compare(unit) == 0)
    return true;
  if (cm_Per_s.GetString().compare(unit) == 0)
    return true;
  if (m_Per_min.GetString().compare(unit) == 0)
    return true;
  if (cm_Per_min.GetString().compare(unit) == 0)
    return true;
  if (ft_Per_s.GetString().compare(unit) == 0)
    return true;
  if (ft_Per_min.GetString().compare(unit) == 0)
    return true;
  if (km_Per_hr.GetString().compare(unit) == 0)
    return true;
  return false;
}

const LengthPerTimeUnit& LengthPerTimeUnit::GetCompoundUnit(const std::string& unit)
{
  if (m_Per_s.GetString().compare(unit) == 0)
    return m_Per_s;
  if (cm_Per_s.GetString().compare(unit) == 0)
    return cm_Per_s;
  if (m_Per_min.GetString().compare(unit) == 0)
    return m_Per_min;
  if (cm_Per_min.GetString().compare(unit) == 0)
    return cm_Per_min;
  if (ft_Per_s.GetString().compare(unit) == 0)
    return ft_Per_s;
  if (ft_Per_min.GetString().compare(unit) == 0)
    return ft_Per_min;
  if (km_Per_hr.GetString().compare(unit) == 0)
    return km_Per_hr;
  throw CommonDataModelException(unit + " is not a valid LengthPerTime unit");
}

template class SEScalarQuantity<LengthPerTimeUnit>;
